src/tokens/tokens.service.ts

```diff
@@ -0,0 +1,398 @@
+    this.topic = topic;
+    this.shortPrefix = shortPrefix;
+    this.contractABI = contractABI;
+    this.contractInstanceUrl = `${this.baseUrl}/abis/${contractABI}`;
+    this.username = username;
+    this.password = password;
+    this.proxy.addListener(
+      new TokenListener(this.http, this.instanceUrl, this.topic, this.username, this.password),
+    );
+  }
+
+  /**
+   * One-time initialization of event stream and base subscription.
+   */
+  async init() {
+    this.stream = await this.eventstream.createOrUpdateStream(this.topic);
+    await this.eventstream.getOrCreateSubscription(
+      this.instanceUrl,
+      this.stream.id,
+      tokenCreateEvent,
+      packSubscriptionName(this.topic, BASE_SUBSCRIPTION_NAME),
+    );
+  }
+
+  /**
+   * If there is an existing event stream whose subscriptions don't match the current
+   * events and naming format, delete the stream so we'll start over.
+   * This will cause redelivery of all token events, which will poke FireFly to
+   * (re)activate pools and (re)process all transfers.
+   *
+   * TODO: eventually this migration logic can be pruned
+   */
+  async migrate() {
+    const streams = await this.eventstream.getStreams();
+    const existingStream = streams.find(s => s.name === this.topic);
+    if (existingStream === undefined) {
+      return;
+    }
+    const subscriptions = await this.eventstream.getSubscriptions();
+    if (subscriptions.length === 0) {
+      return;
+    }
+
+    const foundEvents = new Set<string>();
+    for (const sub of subscriptions.filter(s => s.stream === existingStream.id)) {
+      const parts = unpackSubscriptionName(this.topic, sub.name);
+      if (parts.event !== undefined && parts.event !== '') {
+        foundEvents.add(parts.event);
+      }
+    }
+
+    if (foundEvents.size === 1 && foundEvents.has(BASE_SUBSCRIPTION_NAME)) {
+      // Special case - only the base subscription exists (with the correct name),
+      // but no pools have been activated. This is ok.
+      return;
+    }
+
+    // Otherwise, expect to have found subscriptions for each of the events.
+    for (const event of ALL_SUBSCRIBED_EVENTS) {
+      if (!foundEvents.has(event)) {
+        this.logger.warn('Incorrect event stream subscriptions found - deleting and recreating');
+        await this.eventstream.deleteStream(existingStream.id);
+        await this.init();
+        break;
+      }
+    }
+  }
+
+  private postOptions(operator: string, requestId?: string) {
+    const from = `${this.shortPrefix}-from`;
+    const sync = `${this.shortPrefix}-sync`;
+    const id = `${this.shortPrefix}-id`;
+
+    const requestOptions: AxiosRequestConfig = {
+      params: {
+        [from]: operator,
+        [sync]: 'false',
+        [id]: requestId,
+      },
+      ...basicAuth(this.username, this.password),
+    };
+
+    return requestOptions;
+  }
+
+  async createPool(dto: TokenPool): Promise<AsyncResponse> {
+    const response = await lastValueFrom(
```

Comment:
You should verify that `dto.type` is fungible, and throw an HTTP "bad request" if it's not.